### ISSUE TITLE
Issue with fix missing spaces in Arabic

### ISSUE THREAD

**Issue by OmrSi** [SubtitleEdit/subtitleedit#4077]
I have noticed many subtitles with issues in spaces before the Arabic comma and question mark, when I asked the uploaders of said subtitles, they said they use "Fix missing spaces" in SE.
So I tested and found this:
[Test.zip](https://github.com/SubtitleEdit/subtitleedit/files/4410140/Test.zip)

SE should not add space before the Arabic comma and question mark.
On the contrary, it must remove any spaces before them.
The Arabic comma "،" and question mark "؟" never comes with a space before them.
It should be removed from "Fix missing spaces" and added to "Remove unneeded spaces".
BTW, we use `()` and `""` for names when translating to Arabic to distinguish names from normal words.

Here is another case I forgot to mention: `ما الأخبار أيها الـ"اسم"؟`
There shouldn't be a space after `ـ` and it should be deleted if it exists.

Another one: `و"الشيء"`
There shouldn't be a space after a single "و" and it should be removed if it exists.

To sum it up, spaces before commas and question marks must be removed.
Spaces after `ـ` and single `و` must be remvoed.

**Comment by niksedk** [comment]
Sorry, is "after" in Arabic like "before" in English?

![image](https://user-images.githubusercontent.com/1261331/78051626-404bbb80-737e-11ea-91d4-c2cd67ebc318.png)

Is this code correct?
```java
while (text.Contains("؟ "))
{
       text = text.Replace("؟ ", "؟");
}

while (text.Contains("\u060C ")) // Arabic comma
{
       text = text.Replace("\u060C ", "\u060C");
}

while (text.Contains(" و "))
{
       text = text.Replace(" و ", "و ");
}

while (text.Contains(" ـ "))
{
       text = text.Replace(" ـ ", "ـ ");
}
```

**Comment by OmrSi** [comment]
It's the other way around, from the looks of it.
Space before the comma is "space, comma", " ،"
Space before the question mark is "space, question mark", " ؟"
Space after "ـ" is `ـ `
Space after "و" is `و `
Note that the "و" must be alone, not connected to anything before it, so "نجو" doesn't count.

**Comment by niksedk** [comment]
How is latest master now?

**Comment by OmrSi** [comment]
@niksedk Still buggy.
So, "ـ" could be connected to something before it like `الـ"ولد"`
Also, the `و` can have a quote before it, when a line is on-screen, you put it in quotes: `"و كان الرجل وسيمًا"`
Plus, "Fix missing spaces" is still adding these things, if "Fix missing spaces" is active, it still adds the spaces to the wrong place.

**Comment by niksedk** [comment]
OK, updated "Remove unneeded spaces" again.

About the "Fix missing spaces"... what are the rules for Arabic? Space after comma, period and question-mark?

**Comment by OmrSi** [comment]
Yep, except if the are in a quote like: `"How are you?"`

**Comment by niksedk** [comment]
thx - I've now guessed a bit and tried to add Arabic support to "Fix missing spaces"... 
How does it work?

**Comment by OmrSi** [comment]
It's working well now, as far as I can see. :D

Just two things:
This case: `الـ"ولد"` is still seen as missing space, and this case: `و"الولد` too, and it's not deleting the space if it exists.
As I said, `و` can be alone or preceded by a quote or the beginning of the line, while "ـ" comes always connected to something before it and never has a space after it. If there is a space, it should be deleted, if there isn't, it shouldn't be added.

The second thing is with "Remove lines breaks in all lines the aren't dialog".
SE isn't detecting lines the end with `؟` as dialog, like:
```
- من هذا؟
- (جينا).
```

**Comment by niksedk** [comment]
The first issue I cannot re-create with "fix missing spaces" checked and "Arabic" as language:

![image](https://user-images.githubusercontent.com/1261331/78178806-7d887a00-7460-11ea-88e6-efa15db7480a.png)

![image](https://user-images.githubusercontent.com/1261331/78178724-621d6f00-7460-11ea-979b-01332d53df74.png)

**Comment by OmrSi** [comment]
Sorry, that was my fault, I didn't choose the language as it's set to skip the list.
It works fine with Arabic, thank you.
But it's still not removing the space in:
`و الـ "ولد"`
After the single "و" at the begging of the line and after the "ـ".

**Comment by niksedk** [comment]
thx for testing - could you re-test?

**Comment by OmrSi** [comment]
Yep. Works well. Thank you.
The only thing left is that it's not detecting "؟" as an end of line for dialog in "Merge short lines except for dialog" as in:
```
- من هذا؟
- (جينا).
```

**Comment by OmrSi** [comment]
@niksedk To fix this, you just need to add `؟` to the ifs in lines number: 511 and 531 of `libse/Utilities.cs`

**Comment by OmrSi** [comment]
@niksedk I found something wrong.
It's removing the spaces before the "و" in lines like:
`"بيبي" و"دانيت".`
There are two names, the "و" is stuck to what's after it, it shouldn't remove the space before it.

**Comment by niksedk** [comment]
Thx for testing :)
Better ?

**Comment by OmrSi** [comment]
Yep. Thank you. <3

**Comment by OmrSi** [comment]
Since it's related to this...
@niksedk This shouldn't change in "Fix missing spaces":
![image](https://user-images.githubusercontent.com/20923700/82857222-11cf0480-9f19-11ea-9189-7726608387e6.png)

It's: `...ذهب إلى هناك.`
That is only at the beginning of a line.

**Comment by OmrSi** [comment]
@niksedk When you use "Reverse RTL Start/End" the dot at the end moves to the beginning and so SE sees it as a dot with no space after it and adds a space, can you make it ignore that case as well?
Like: ".ذهب إلى هناك"
This is reversed and a space shouldn't be added.
